### PR TITLE
make SizeLimit a trait

### LINKED ISSUE #121: Consider making SizeLimit a trait rather than an enum
```diff
-pub struct Deserializer<R> {
+pub struct Deserializer<R, L> {
     reader: R,
-    size_limit: SizeLimit,
+    size_limit: L,
     read: u64
 }
```

This way the Infinite case would not be slowed down by all the branching in Deserializer::read_bytes:

```rust
match self.size_limit {
    SizeLimit::Infinite => Ok(()),
    SizeLimit::Bounded(x) if self.read <= x => Ok(()),
    SizeLimit::Bounded(_) => Err(ErrorKind::SizeLimit.into())
}
```

Something like:

```rust
trait SizeLimit {
    fn check(&self, read: u64) -> Result<()>;
}

struct Infinite;
impl SizeLimit for Infinite {
    #[inline]
    fn check(&self, _read: u64) -> Result<()> {
        Ok(())
    }
}

struct Bounded(u64);
impl SizeLimit for Bounded {
    fn check(&self, read: u64) -> Result<()> {
        if read <= self.0 {
            Ok(())
        } else {
            Err(ErrorKind::SizeLimit.into())
        }
    }
}

impl<R: Read, L: SizeLimit> Deserializer<R, L> {
    fn read_bytes(&mut self, count: u64) -> Result<()> {
        self.read += count;
        self.size_limit.check(self.read)
    }
}
```

The compiler may be able to generate faster code in the Infinite case knowing that no branching is happening.

### PR DESCRIPTION
**Pull request by TyOverby**
Closes #121

### CODE REVIEW

**Comment by TyOverby** [comment]
cc @dtolnay

**Comment by dtolnay** [comment]
This didn't make much of a difference in #123...

#### Before

```
test bincode_deserialize ... bench:          42 ns/iter (+/- 0)
test bincode_serialize   ... bench:          82 ns/iter (+/- 1)
```

#### After

```
test bincode_deserialize ... bench:          36 ns/iter (+/- 1)
test bincode_serialize   ... bench:          84 ns/iter (+/- 4)
```

I don't have time to review the code tonight but maybe read through the code again and focus on places where the Infinite / &[u8] case has different behavior compared to my implementation.

File: src/lib.rs [hunk 1]
```diff
@@ -119,8 +118,32 @@ pub fn serialize<T: ?Sized>(value: &T, size_limit: SizeLimit) -> serde::Result<V
 /// encoding function, the encoder will verify that the structure can be encoded
 /// within that limit.  This verification occurs before any bytes are written to
 /// the Writer, so recovering from an error is easy.
-#[derive(Clone, Copy, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
-pub enum SizeLimit {
-    Infinite,
-    Bounded(u64)
+pub trait SizeLimit {
+    #[inline(always)]
```
Comment (dtolnay) [inline src/lib.rs]: My understanding is this doesn't do anything, the compiler just ignores it. The inline attribute needs to go where the body of the method is defined.

Comment (TyOverby) [inline src/lib.rs]: Oh dang.

**Comment by TyOverby** [comment]
the serde-rs/bench test is a *really* micro-benchmark.

I think that I'm going to spend some time and make a benchmarking system that has a larger variety of tests.

**Pull request merged**